internal/engine/event_watch_manager_test.go

```diff
@@ -0,0 +1,131 @@
+package engine
+
+import (
+	"context"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/windmilleng/tilt/internal/k8s"
+	"github.com/windmilleng/tilt/internal/model"
+	"github.com/windmilleng/tilt/internal/store"
+	"github.com/windmilleng/tilt/internal/testutils/output"
+	v1 "k8s.io/api/core/v1"
+)
+
+func TestEventWatchManager_dispatchesEvent(t *testing.T) {
+	f := newEWMFixture(t)
+	defer f.TearDown()
+
+	f.addManifest("someK8sManifest")
+
+	evt := &v1.Event{
+		Reason:  "because test",
+		Message: "hello world",
+	}
+
+	f.ewm.OnChange(f.ctx, f.store)
+	f.kClient.EmitEvent(f.ctx, evt)
+	expectedAction := store.K8SEventAction{Event: evt}
+	f.assertObservedK8sEventActions(expectedAction)
+
+}
+func TestEventWatchManager_needsWatchNoK8s(t *testing.T) {
+	f := newEWMFixture(t)
+	defer f.TearDown()
+
+	evt := &v1.Event{
+		Reason:  "because test",
+		Message: "hello world",
+	}
+
+	// No k8s manifests on the state, so OnChange shouldn't do anything --
+	// when we emit an event, we do NOT expect to see an action dispatched,
+	// since no watch should have been started.
+	f.ewm.OnChange(f.ctx, f.store)
+	f.kClient.EmitEvent(f.ctx, evt)
+	f.assertNoK8sEventActions()
+}
+
+type ewmFixture struct {
+	t          *testing.T
+	kClient    *k8s.FakeK8sClient
+	ewm        *EventWatchManager
+	ctx        context.Context
+	cancel     func()
+	store      *store.Store
+	getActions func() []store.Action
+
+	// old value of k8sEventsFeatureFlag env var, for teardown
+	// TODO(maia): remove this when we remove the feature flag
+	oldFeatureFlagVal string
+}
+
+func newEWMFixture(t *testing.T) *ewmFixture {
+	kClient := k8s.NewFakeK8sClient()
+
+	ctx := output.CtxForTest()
+	ctx, cancel := context.WithCancel(ctx)
+
+	ret := &ewmFixture{
+		kClient:           kClient,
+		ewm:               NewEventWatchManager(kClient),
+		ctx:               ctx,
+		cancel:            cancel,
+		t:                 t,
+		oldFeatureFlagVal: os.Getenv(k8sEventsFeatureFlag),
+	}
+
+	os.Setenv(k8sEventsFeatureFlag, "true")
+
+	ret.store, ret.getActions = store.NewStoreForTesting()
+	go ret.store.Loop(ctx)
+
+	return ret
+}
+
+func (f *ewmFixture) TearDown() {
+	_ = os.Setenv(k8sEventsFeatureFlag, f.oldFeatureFlagVal)
+	f.cancel()
+}
+
+func (f *ewmFixture) addManifest(manifestName string) {
+	state := f.store.LockMutableStateForTesting()
+	state.WatchFiles = true
+	dt := model.K8sTarget{Name: model.TargetName(manifestName)}
+	m := model.Manifest{Name: model.ManifestName(manifestName)}.WithDeployTarget(dt)
+	state.UpsertManifestTarget(store.NewManifestTarget(m))
+	f.store.UnlockMutableState()
+}
+
+func (f *ewmFixture) assertNoK8sEventActions() {
+	f.assertObservedK8sEventActions()
+}
+
+func (f *ewmFixture) assertObservedK8sEventActions(expectedActions ...store.K8SEventAction) {
+	if len(expectedActions) == 0 {
+		// assert no k8s event actions -- sleep briefly
+		// to give any actions a chance to get into the queue
+		time.Sleep(10 * time.Millisecond)
```

Comment:
should this sleep be after the for loop, to catch the case where you get too many actions?

Reply:
...wait what?

so the thinking is that if we check immediately and get no actions, that doesn't necessarily mean anything cuz maybe the actions took a few ms to get into the queue. we wait a few ms so that we can be more certain that we ACTUALLY have no actions, rather than that we just checked too fast and the actions haven't show up yet.

Reply:
ya, i understand that. i'm saying that
 "assert 0 actions but have to wait a few ms to see if any more actions get in the queue" 
is a special case of
"assert N actions but have to wait a few ms to see if N + 1 actions get in the queue"
and this assertion would be simpler and better if it handled the general case.

Reply:
OH! yes you're right, good call

Reply:
...though then we wait 200ms every time we assert 0 actions, rather than waiting just 10 ms... i might bump that timeout down abit.